### PR TITLE
DPLY-40 Using DbSessionFactory and DbSession to bridge the Hibernate 3 vs. 4 differences

### PR DESCRIPTION
**Pull request by HerbertYiga**
For this,i am wondering on upgrading core,i successfully introduce DbSession and DbSessionFactory,but  but i run into The method setPrecise(boolean) is undefined for the type ConceptNumeric!!! this the ticket link  https://issues.openmrs.org/browse/DPLY-40?filter=-1   The orignal commit for this ticket were i made the changes is https://github.com/openmrs/openmrs-module-metadatadeploy/commit/53278ecc2150d896644396d0de03329db7a0ebd5

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 1]
```diff
@@ -101,6 +101,7 @@ public void testCreateNumeric() throws Exception {
         numeric.addDescription(new ConceptDescription("1-10", Locale.ENGLISH));
         numeric.setLowAbsolute(1d);
         numeric.setHiAbsolute(10d);
+        //The method setPrecise(boolean) is undefined for the type ConceptNumeric
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Because it was removed at https://github.com/openmrs/openmrs-core/commit/d65825b685453515f55e30504e9cd92ebaa3f759#diff-c6944a72a4cb885dba94e8b377659090 prior openmrCore 2.1.0

File: pom.xml [hunk 2]
```diff
@@ -40,8 +40,8 @@
 	</modules>
 	
 	<properties>
-		<openMRSVersion>1.9.4</openMRSVersion>
-		<openMRSBuildVersion>1.9.4</openMRSBuildVersion>
+		<openMRSVersion>2.1.0</openMRSVersion>
```
Comment (dkayiwa) [inline pom.xml]: Why are you switching to OpenMRS version 2.1.0? Can't you just use 1.9.9? https://wiki.openmrs.org/display/docs/Supporting+Platform+2.0+and+below

Comment (HerbertYiga) [reply pom.xml]: i successfully introduced the 1.9.9 version ,i go on with the work in PersonAttributeTypeDeployHandler.java as above where i introduced the DbSessionFactory and DbSession.But on building i get a DbSessionFactory class not found error? could i be missing out some thing in my code

Comment (dkayiwa) [reply pom.xml]: have you committed?

Comment (HerbertYiga) [reply pom.xml]: yes i have https://github.com/openmrs/openmrs-module-metadatadeploy/pull/8/commits/caeebe2b28c058a2fa4c5ddb13d078b16940f2dc

Comment (HerbertYiga) [reply pom.xml]: yes i have committed!!  https://github.com/openmrs/openmrs-module-metadatadeploy/pull/8/commits/caeebe2b28c058a2fa4c5ddb13d078b16940f2dc

Comment (dkayiwa) [reply pom.xml]: If you have committed, then travis should show you the current error log

Comment (HerbertYiga) [reply pom.xml]: according to travis i get (cannot  find symbol method setPrecise) as seen here https://pastebin.com/WVVnNEvX   now i am trying on another version of core as per the link you dropped me.

Comment (HerbertYiga) [reply pom.xml]: it still puts out (java.lang.ClassNotFoundException: org.openmrs.api.db.hibernate.DbSessionFactory) yet i successfully import this class!!

Comment (dkayiwa) [reply pom.xml]: Is that reported by travis?

Comment (HerbertYiga) [reply pom.xml]: travis is saying  (cannot find symbol method setPrecise) as seen here  https://pastebin.com/WVVnNEvX

Comment (HerbertYiga) [reply pom.xml]: i am just wondering could i be missing out some thing during the importaion.I  get  (java.lang.ClassNotFoundException: org.openmrs.api.db.hibernate.DbSessionFactory).Yet on the importation,i successfully import the DbSessionFactory

Comment (dkayiwa) [reply pom.xml]: Does this pull request have your changes?

Comment (HerbertYiga) [reply pom.xml]: yes it does in the .java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java

Comment (dkayiwa) [reply pom.xml]: Then why doesn't travis report the same error?

Comment (HerbertYiga) [reply pom.xml]: let me commit again!

**Comment by dkayiwa** [comment]
The commit message should include the ticket id as stated at https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by HerbertYiga** [comment]
i have included the ticket ID in the commit message

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 1 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Is the above comment really required?

**Comment by samuelmale** [comment]
Can you also squash your commits into one?  https://wiki.openmrs.org/display/docs/Pull+Request+Tips

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/RoleDeployHandlerTest.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import org.openmrs.Privilege;
 import org.openmrs.Role;
 import org.openmrs.api.context.Context;
+import org.openmrs.api.db.hibernate.DbSessionFactory;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/RoleDeployHandlerTest.java]: Can you remove the `SessionFactory` import, I believe it's nolonger used?

File: api/src/main/java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java [hunk 4]
```diff
@@ -16,10 +16,13 @@
 
 import java.lang.reflect.Method;
 
-import org.hibernate.SessionFactory;
+
+
+
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java]: Why add all these new lines?

File: api/src/main/java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java [hunk 5]
```diff
@@ -80,14 +79,14 @@ public void uninstall(PersonAttributeType obj, String reason) {
 	 * 
 	 * @return the current hibernate session.
 	 */
-	private org.hibernate.Session getCurrentSession() {
+	private  org.openmrs.api.db.hibernate.DbSession getCurrentSession() {
```
Comment (samuelmale) [inline api/src/main/java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java]: Can you correct your indention above?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 6]
```diff
@@ -56,223 +56,192 @@
 	@Autowired
 	private MetadataDeployService deployService;
 
-    @Autowired
-    private ConceptService conceptService;
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Why did you remove this here?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 7]
```diff
@@ -56,223 +56,192 @@
...
-                hasProperty("localePreferred", is(true)),
-                hasProperty("conceptNameType", is(ConceptNameType.FULLY_SPECIFIED)),
-                hasProperty("tags", containsInAnyOrder(Arrays.asList(
-                                hasProperty("tag", is("refapp-preferred"))
-                        )
-                ))));
-        assertThat(created.getDescriptions().size(), is(1));
-        assertThat(created.getDescription(), allOf(
-                hasProperty("description", is("How the patient describes it")),
-                hasProperty("locale", is(Locale.ENGLISH))));
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 8]
```diff
@@ -56,223 +56,192 @@
...
-    }
-
-    @Test
-    public void testCreateNumeric() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 9]
```diff
@@ -56,223 +56,192 @@
...
-
-        ConceptNumeric created = (ConceptNumeric) conceptService.getConceptByUuid(UUID);
-        assertThat(created.getLowAbsolute(), is(1d));
-        assertThat(created.getHiAbsolute(), is(10d));
-        assertThat(created.getPrecise(), is(false));
-        assertThat(created.getName().getName(), is("Favorite single digit number"));
-    }
-
-    @Test
-    public void testCreateNumericUsingBuilder() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 10]
```diff
@@ -56,223 +56,192 @@
...
-
-        ConceptNumeric created = (ConceptNumeric) conceptService.getConceptByUuid(UUID);
-        assertThat(created.getLowAbsolute(), is(1d));
-        assertThat(created.getHiAbsolute(), is(10d));
-        assertThat(created.getPrecise(), is(false));
-        assertThat(created.getName().getName(), is("Favorite single digit number"));
-    }
-
-    @Test
-    public void testUpdate() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 11]
```diff
@@ -56,223 +56,192 @@
...
-                                hasProperty("conceptSource", hasProperty("uuid", is("j3nfjk33-639f-4cb4-961f-1e025b908433")))
-                        ))
-                )));
-
-        // Check everything can be persisted
-        Context.flushSession();
-    }
-
-    @Test
-    public void testUpdateNumeric() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 12]
```diff
@@ -56,223 +56,192 @@
...
-
-        ConceptNumeric updated = (ConceptNumeric) conceptService.getConceptByUuid(weightUuid);
-        assertThat(updated.getLowAbsolute(), is(0d));
-        assertThat(updated.getHiAbsolute(), nullValue());
-        assertThat(updated.getPrecise(), is(true));
-        assertThat(updated.getName().getName(), is("WEIGHT (KG)"));
-    }
-
-    @Test
-    public void testRetire() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 13]
```diff
@@ -56,223 +56,192 @@
...
-        executeDataSet("concept/existingConcepts.xml");
-        deployService.uninstallObject(deployService.fetchObject(Concept.class, UUID), "Testing");
-        assertThat(Context.getConceptService().getConceptByUuid(UUID).isRetired(), is(true));
-    }
-
-    @Test
-    public void testUnretire() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java [hunk 14]
```diff
@@ -56,223 +56,192 @@
...
-
-        Concept unretired = conceptService.getConcept(1790);
-        assertThat(unretired.isRetired(), is(false));
-        assertThat(unretired.getDateRetired(), is(nullValue()));
-        assertThat(unretired.getRetiredBy(), is(nullValue()));
-        assertThat(unretired.getRetireReason(), is(nullValue()));
-    }
-
-    @Test
-    public void testOverwriteSet() throws Exception {
```
Comment (samuelmale) [inline api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: Did you make any changes in this method? If so, did you rewrite the whole method?

**Comment by dkayiwa** [comment]
@HerbertYiga can you respond to each of @samuelmale's comments on this ticket?

File: api/src/main/java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java (hunk 5 above)
Comment (HerbertYiga) [reply api/src/main/java/org/openmrs/module/metadatadeploy/handler/impl/PersonAttributeTypeDeployHandler.java]: i have corrected that

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 6 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not do any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 7 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not do any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 8 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 9 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 10 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 11 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 12 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 13 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

File: api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java (hunk 14 above)
Comment (HerbertYiga) [reply api/src/test/java/org/openmrs/module/metadatadeploy/handler/impl/ConceptDeployHandlerTest.java]: i did not make any changes here and have removed the file that i had pushed

**Comment by HerbertYiga** [comment]
@dkayiwa  i have responded to all  @samuelmale's comments as suggested.

**Comment by dkayiwa** [comment]
Can you include a link to the ticket as mentioned at? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by HerbertYiga** [comment]
@dkayiwa  i have edited my first comment and included the ticket link as suggested.

File: api-1.10/pom.xml [hunk 15]
```diff
@@ -18,7 +18,7 @@
         <MODULE_NAME>${project.parent.name}</MODULE_NAME>
         <MODULE_VERSION>${project.parent.version}</MODULE_VERSION>
         <MODULE_PACKAGE>${project.parent.groupId}.${project.parent.artifactId}</MODULE_PACKAGE>
-        <openMRSBuildVersion>1.10.0</openMRSBuildVersion>
```
Comment (dkayiwa) [inline api-1.10/pom.xml]: Is there any reason why you switched from the 1.10.x series to the 1.11.x series?

Comment (HerbertYiga) [reply api-1.10/pom.xml]: yes.You advised me to use 1.11x series so as to avail DbSession and DbSessionFactory.

Comment (dkayiwa) [reply api-1.10/pom.xml]: No i did not say that.

Comment (HerbertYiga) [reply api-1.10/pom.xml]: okay.I tried 1.10.x series,DbSession and DbSessionFactory were not available so i had to use 1.11.x series

Comment (dkayiwa) [reply api-1.10/pom.xml]: Which exact version did you try?

Comment (HerbertYiga) [reply api-1.10/pom.xml]: i resumed trying on 1.10.x series versions and got 1.10.4 having a success build,its what i have commited

**Comment by dkayiwa** [comment]
As indicated in the ticket description, this ticket also requires a revert of the original commit linked in the ticket description.

**Comment by HerbertYiga** [comment]
@dkayiwa   I have included the link for the orignal commit in my ticket  description.

**Comment by dkayiwa** [comment]
Did you understand my comment? Did you also understand the ticket description?

**Comment by HerbertYiga** [comment]
@dkayiwa when  i saw"Revert of the orignal commit", i got confused a bit but i first wanted to try out my answer before getting the clearer meaning thats why i included the link.could i be missing out some thing?

**Comment by dkayiwa** [comment]
Did you read about the ticket description saying that the original commit was not the preferred approach?

**Comment by HerbertYiga** [comment]
@dkayiwa Not at all beacuse this is what i have in my ticket description "Description
In response to this https://talk.openmrs.org/t/metadata-deploy-bug/10187
A commit was made at https://github.com/openmrs/openmrs-module-metadatadeploy/commit/53278ecc2150d896644396d0de03329db7a0ebd5

This ticket is about changing the above commit to use the DbSessionFactory and DbSession classes as recommended at https://wiki.openmrs.org/display/docs/Supporting+Platform+2.0+and+below"

**Comment by HerbertYiga** [comment]
@ didnot get any thing about the approach

**Comment by dkayiwa** [comment]
On that wiki page, did you read this?

For module developers who do not want to require upgrading to the latest maintenance version of the platform, you could use an alternative approach as done by the xforms module at https://github.com/openmrs/openmrs-module-xforms/commit/5d79f327d27777a40ec755a90433e4dc69067e4a

**Comment by HerbertYiga** [comment]
@dkayiwa when i look at that link it looks like i have to remove the DbSession and DbSessionFactory!! could i be following the right approach??

**Comment by dkayiwa** [comment]
You need to revert the entire original commit

**Comment by HerbertYiga** [comment]
@dkayiwa am still abit confused,do i have to return to my initial commit and i change the approach,some clarification on this!!

**Comment by dkayiwa** [comment]
You need to also include a reverse of whatever is in that commit.

**Comment by HerbertYiga** [comment]
according to this link, https://github.com/openmrs/openmrs-module-xforms/commit/5d79f327d27777a40ec755a90433e4dc69067e4a that you droped me,it looks like i donot have to make changes for this ticket since it cleary handles the creating a sessionFactory without upgrading core!but incase i am to use DbSession and DbSessionFactory,then i have to upgrade core

**Comment by dkayiwa** [comment]
Let me put it this way, for every green change in the original commit, there should be an equivalent red change in your pull request. That is what we call a revert.

**Comment by HerbertYiga** [comment]
okie fine i now get the picture but am wondering whether the code i commited initially was fine becuase you said i was using a wrong approach and according to the approach you sent me it looks like i have to change nothing

**Comment by dkayiwa** [comment]
We cannot tell if your changes work fine on their own until you include a revert of the original changes.

**Comment by HerbertYiga** [comment]
okie a fine am going to make a new commit,i googled on git revert,it was though telling me that it helps to make a step forward.though i thought the red and green just comes automatically!!
